File: delivery/be/document_detail.py
"""Read-only document detail and stored translation projection."""
from __future__ import annotations

from delivery.be.catalogue import load_taxonomy


def collect_document_detail(conn, seq_id: int, taxonomy: dict | None = None) -> dict | None:
    taxonomy = taxonomy if taxonomy is not None else load_taxonomy()
    row = conn.execute(
        """
        SELECT d.*, s.site_name, s.site_url, s.sheet, COALESCE(dl.lang, 'unknown') AS lang
        FROM documents d JOIN sites s USING (site_id)
        LEFT JOIN document_lang dl USING (seq_id)
        WHERE d.seq_id=%s
        """,
        (seq_id,),
    ).fetchone()
    if row is None:
        return None
    translations = conn.execute(
        """
        SELECT DISTINCT ON (source_field) source_field, translation_text, model_version,
               prompt_version, completed_at
        FROM document_translations
        WHERE seq_id=%s AND target_locale='ko-KR' AND state='completed'
          AND source_field IN ('title', 'description')
        ORDER BY source_field, completed_at DESC NULLS LAST, updated_at DESC, translation_id DESC
        """,
        (seq_id,),
    ).fetchall()
    translated = {
        item["source_field"]: {
            "text": item["translation_text"], "model_version": item["model_version"],
            "prompt_version": item["prompt_version"], "completed_at": item["completed_at"],
        }
        for item in translations
    }
    summary = conn.execute(
        """SELECT summary_text,key_points,institutions,source_facts,model_version,prompt_version,completed_at
           FROM document_summaries WHERE seq_id=%s AND target_locale='ko-KR' AND state='completed'
           ORDER BY completed_at DESC NULLS LAST,updated_at DESC,summary_id DESC LIMIT 1""",
        (seq_id,),
    ).fetchone()
    classification = taxonomy.get(row["site_id"], {"country": "기타", "doc_type": "기타"})
    return {
        "seq_id": row["seq_id"],
        "site": {key: row[key] for key in ("site_id", "site_name", "site_url", "sheet")},
        "classification": classification,
        "source": {
            key: row[key] for key in (
                "lang", "title", "abstract", "authors", "publisher", "journal", "keywords",
                "published_date", "listed_date", "collected_at", "summary", "summary_model",
                "meta_url", "pdf_url",
            )
        },
        "translations": {
            "target_locale": "ko-KR", "title": translated.get("title"),
            "description": translated.get("description"),
        },
        "generated_summary": dict(summary) if summary else None,
        "files": {
            "has_pdf": bool(row["pdf_downloaded"]), "has_text": bool(row["text_extracted"]),
            "pdf_size_bytes": row["pdf_size_bytes"], "original_filename": row["original_filename"],
        },
    }

File: delivery/fe/src/lib/document-detail.ts
import "server-only";

type Translation = Readonly<{
  text: string;
  model_version: string;
  prompt_version: string;
  completed_at: string | null;
}>;

export type DocumentDetail = Readonly<{
  seq_id: number;
  site: Readonly<{ site_id: string; site_name: string; site_url: string; sheet: string | null }>;
  classification: Readonly<{ country: string; doc_type: string }>;
  source: Readonly<{
    lang: string; title: string; abstract: string | null; authors: string | null;
    publisher: string | null; journal: string | null; keywords: string | null;
    published_date: string | null; listed_date: string | null; collected_at: string | null;
    summary: string | null; summary_model: string | null; meta_url: string; pdf_url: string | null;
  }>;
  translations: Readonly<{ target_locale: string; title: Translation | null; description: Translation | null }>;
  generated_summary: Readonly<{ summary_text:string; key_points:string[]; institutions:string[];
    source_facts:Readonly<{urls?:string[];dates?:string[];numbers?:string[]}>;model_version:string;
    prompt_version:string;completed_at:string|null }> | null;
  files: Readonly<{ has_pdf: boolean; has_text: boolean; pdf_size_bytes: number | null; original_filename: string | null }>;
}>;

export type DetailResult =
  | Readonly<{ ok: true; data: DocumentDetail }>
  | Readonly<{ ok: false; kind: "not_found" | "unavailable" }>;

const backendUrl = (): string => (process.env.BE_URL ?? "http://127.0.0.1:8080").replace(/\/$/, "");

export async function getDocumentDetail(seqId: string): Promise<DetailResult> {
  try {
    const response = await fetch(`${backendUrl()}/documents/${encodeURIComponent(seqId)}`, {
      cache: "no-store", signal: AbortSignal.timeout(8000),
    });
    if (response.status === 404 || response.status === 422) return { ok: false, kind: "not_found" };
    if (!response.ok) return { ok: false, kind: "unavailable" };
    return { ok: true, data: await response.json() as DocumentDetail };
  } catch {
    return { ok: false, kind: "unavailable" };
  }
}
